### PR TITLE
Set default yaw axis Vector3::UNIT_Y (tabletop mode) + Optional X, Z and free rotation

### PR DESCRIPTION
**Pull request by sercero**
This solves #18

### CODE REVIEW

**Comment by sercero** [comment]
Perhaps it would be better if this was an option... for me this is the most reasonable mode

**Comment by paroj** [comment]
I had this on initially. However, this assumes the model actually has y-up. If this is not the case, there is no way to properly orient the model.

An alternative would be to make this configurable:
- free rotation
- fixed x-up
- fixed y-up
- fixed z-up

**Comment by sercero** [comment]
Is there a way to detect if the model does not have Y-up and rotate everything?

Perhaps that would be faster.

In general one would open models that were exported for OGRE which is Y-up, so it makes sense that the default is this.

I think it would be easier to make this selectable and `True` by default.

If someone loads a model that is not Y-up (the rare case) then they can choose to disable this.

**Comment by paroj** [comment]
> In general one would open models that were exported for OGRE which is Y-up, so it makes sense that the default is this.

like I described here https://github.com/OGRECave/blender2ogre/issues/240 it is trivial for user code to change which axis is "up". So I would not assume Y-up.

Also, since assimp is used, you are not even limited to ogre specific files, but maybe just want to inspect how a .obj would look when imported directly in Ogre. There any up assumptions goes out of the window.

> Is there a way to detect if the model does not have Y-up and rotate everything?

one could look at the extents, but you need to assume whether the object is a statue or a bench. so generally: no.

**Comment by sercero** [comment]
> I had this on initially. However, this assumes the model actually has y-up. If this is not the case, there is no way to properly orient the model.
> 
> An alternative would be to make this configurable:
> 
> * free rotation
> * fixed x-up
> * fixed y-up
> * fixed z-up

The issue is that currently you only have the options:
 * free rotation
 * fixed y-up

Otherwise a modification of OGRE Bites CameraMan would be needed.
Unless we can change this with `setPivotOffset()`

**Comment by sercero** [comment]
```
Traceback (most recent call last):
  File "D:\OGRE2\Built\ogre-meshviewer_24.02-win64\ogre_mesh_viewer.py", line 714, in <module>
    app.initApp()
  File "D:\OGRE2\Built\ogre-meshviewer_24.02-win64\ogre_mesh_viewer.py", line 680, in setup
    camnode.setFixedYawAxis(True, Ogre.Vector3.UNIT_Z);
AttributeError: type object 'Ogre.Ogre.Vector3' has no attribute 'UNIT_Z'
```

Can you help with this?

I can use `(0,0,1)` but `UNIT_Z` would be better.

**Comment by paroj** [comment]
use (0,0,1).
Maintaining UNIT_Z is a PITA, as Vector3 is a typdef of Vector<N, float> and it only makes sense for N==3

File: ogre_mesh_viewer.py [hunk 1]
```diff
@@ -250,6 +274,25 @@ def preRenderTargetUpdate(self, evt):
                 if ImGui.MenuItem("Wireframe Mode", "W", app.cam.getPolygonMode() == Ogre.PM_WIREFRAME):
                     self.app._toggle_wireframe_mode()
 
+                if ImGui.BeginMenu("MeshViewer Orientation"):
```
Comment (paroj) [inline ogre_mesh_viewer.py:277]:
```suggestion
                if ImGui.BeginMenu("Fixed Camera Yaw"):
```

File: ogre_mesh_viewer.py [hunk 2]
```diff
@@ -250,6 +274,25 @@ def preRenderTargetUpdate(self, evt):
...
+                    if ImGui.MenuItem("Free rotation", "", not self.fixed_yaw):
```
Comment (paroj) [inline ogre_mesh_viewer.py:278]:
```suggestion
                    if ImGui.MenuItem("Disabled", "", not self.fixed_yaw):
```

File: ogre_mesh_viewer.py [hunk 3]
```diff
@@ -250,6 +274,25 @@ def preRenderTargetUpdate(self, evt):
...
+                        self.fixed_yaw = False
+                        self.set_orientation()
+                    ImGui.Separator()
+                    if ImGui.MenuItem("Fixed X-up", "", self.fixed_yaw and self.fixed_yaw_axis is 'X-up'):
```
Comment (paroj) [inline ogre_mesh_viewer.py:282]:
```suggestion
                    if ImGui.MenuItem("X axis", "", self.fixed_yaw and self.fixed_yaw_axis is 'X-up'):
```

File: ogre_mesh_viewer.py [hunk 4]
```diff
@@ -201,6 +205,26 @@ def draw_loading(self):
...
+    def set_orientation(self):
+        camnode = self.app.camman.getCamera()
+        diam = camnode.getPosition().length()
+        camnode.setOrientation(Ogre.Quaternion.IDENTITY)
+        if self.fixed_yaw is True:
+            camnode.setFixedYawAxis(True, self.fixed_yaw_axes[self.fixed_yaw_axis])
+            if self.fixed_yaw_axis == 'X-up':
+                self.app.camman.setYawPitchDist(0, 0, diam)
+                camnode.roll(-Ogre.Degree(90))
+                print("X-up")
```
Comment (paroj) [inline ogre_mesh_viewer.py:217]: remove debug output

File: ogre_mesh_viewer.py [hunk 5]
```diff
@@ -139,6 +139,10 @@ def __init__(self, app):
         self.show_render_settings = False
         self.side_panel_visible = True
         self.is_filedialog_open = False
+        self.fixed_yaw_axes = {'X-up': (1, 0, 0), 'Y-up': (0, 1, 0), 'Z-up': (0, 0, 1)}
+        self.fixed_yaw_axis = 'Y-up'
```
Comment (paroj) [inline ogre_mesh_viewer.py:143]:
```suggestion
        self.fixed_yaw_axis = -1
```
where -1 means disabled, 0 means x-axis etc.

then construct the axis as
```py
axis = [0, 0, 0]
axis[self.fixed_yaw_axis] = 1
```

remove self.fixed_yaw

File: ogre_mesh_viewer.py [hunk 6]
```diff
@@ -200,6 +205,23 @@ def draw_loading(self):
         ImGui.Text("Loading..            ")
         ImGui.End()
 
+    def set_orientation(self):
```
Comment (paroj) [inline ogre_mesh_viewer.py:208]: this method only modifes `self.app` therefore it should be moved to `app`

File: ogre_mesh_viewer.py [hunk 7]
```diff
@@ -139,6 +139,11 @@ def __init__(self, app):
         self.show_render_settings = False
         self.side_panel_visible = True
 
+        self.is_filedialog_open = False
```
Comment (paroj) [inline ogre_mesh_viewer.py:142]: no longer used

File: ogre_mesh_viewer.py (hunk 6 above)
Comment (sercero) [reply ogre_mesh_viewer.py:208]: image

**Pull request merged**